frontend/exporter/src/types/copied.rs

```diff
@@ -3302,12 +3330,14 @@ pub struct Binder<T> {
 
 /// Reflects [`rustc_middle::ty::GenericPredicates`]
 #[derive(AdtInto)]
-#[args(<'tcx, S: UnderOwnerState<'tcx>>, from: rustc_middle::ty::GenericPredicates<'tcx>, state: S as tcx)]
+#[args(<'tcx, S: UnderOwnerState<'tcx>>, from: rustc_middle::ty::GenericPredicates<'tcx>, state: S as s)]
 #[derive(
     Clone, Debug, Serialize, Deserialize, JsonSchema, Hash, PartialEq, Eq, PartialOrd, Ord,
 )]
 pub struct GenericPredicates {
     pub parent: Option<DefId>,
+    // FIXME: Switch from `Predicate` to `Clause` (will require correct handling of binders).
+    #[value(self.predicates.iter().map(|(clause, span)| (clause.as_predicate().sinto(s), span.sinto(s))).collect())]
```

Comment:
I'm not sure I understand exactly the renaming done in Rustc.
`rustc_middle::ty::Predicate` and `rustc_middle::ty::Clause` are wrapping exactly the same pieces of information, right? from what I understand it's just a nominal difference, and the two types offer two different sets of methods
But that split seems less relevant in hax, since we provide exactly the same data for those two

Reply:
But then, I don't understand your FIXME comment: both rustc_middle::ty::Predicate and rustc_middle::ty::Clause contains binders of PredicateKind, there should be no difference in term of binders handling

Reply:
I mean, we could just use `Predicate` everywhere, but at the cost of a lot of `=> unreachable!()`.

Reply:
No, what I mean is that there is both `rustc_middle::ty::Predicate` and `rustc_middle::ty::Clause`, which are both newtypes around:

 [Interned](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/intern/struct.Interned.html)<'tcx, [WithCachedTypeInfo](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.WithCachedTypeInfo.html)<[Binder](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.Binder.html)<'tcx, [PredicateKind](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/predicate/type.PredicateKind.html)<'tcx>>>>)

So, this distinction between `Clause` and `Predicate` is purely nominal.

Then, `ClauseKind`, called `Clause` in hax, is indeed a subcase of `PredicateKind`, which itself is a field of `Predicate`.

My point here is that the `clause` in `self.predicates.iter().map(|(clause, span)| (clause.as_predicate(). ....` is of type `rust_middle::ty::Clause`, which is basically exactly a `hax::Predicate`, that is a PredicateKind wrapped into a Binder. So for me, that's not a FIXME, we should just derive a `sinto` for `Clause` that calls `as_predicate` and `sinto` on the result.

Reply:
It's not just nominal, a `Clause` is guaranteed to correspond to a predicate with kind `PredicateKind::Clause`, which is useful because that's fewer cases to handle. Would you prefer we lose this information and use `Predicate` everywhere?

Reply:
Wait, there is *no* formal guarantee about the content of a Clause, there isn't even a smart constructor that would reject a `Clause` that contains something else than a `PredicateKind::Clause`.
The only thing we get is the documentation saying "a subset of predicates which can be assumed by the trait solver"

Do you then mean that we should reflect this informal invariant and define a type `type Clause = Binder<UnboundedClause>`, renaming the current `hax::Clause` into `UnboundedClause`?

Reply:
(well, the API is also quite toward that invariant, since `kind` returns indeed a `Binder<ClauseKind>` -- so having this `Clause` defined as a `Binder<UnboundeClause>` would make sense)

Reply:
What do you mean? The only way to construct a `Clause` is with `Predicate::as_clause` isn't it?

Reply:
The field of `struct Clause(..)` isn't public

Reply:
> Do you then mean that we should reflect this informal invariant and define a type type Clause = Binder<UnboundedClause>, renaming the current hax::Clause into UnboundedClause?

Yep!

Reply:
Yes, you are right, the invariant was not super clear to me, but that's ok indeed.

Alright, then I guess we should just always produce `Binder<ClauseKind>` clauses, never `ClauseKind` by itself; so we should probably move from `Clause` being defined as `ClauseKind + clause id` to `Clause = {kind: Binder<ClauseKind>, id: ClauseId}`. I'm gonna file an issue for that now

Reply:
The problem with that is that we don't have access to the binder when recursively converting `PredicateKind`

Reply:
The thing I had in mind was:
```rust
enum PredicateKind {
  Clause(UnboundClause),
  ...
}
```
with an API to get from `Predicate` to `Clause` by mapping the binder over. Maybe there's a simpler option

Reply:
https://github.com/hacspec/hax/issues/661

Reply:
let's discuss there

Reply:
Resolving this conversation, #661 is a future work